From 010db2eda6957ee678f446b51c6289dd0c81f3b7 Mon Sep 17 00:00:00 2001
From: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
Date: Fri, 18 Nov 2022 07:16:22 -0800
Subject: [PATCH 4/4] ath12k: Optimize tx descriptor updates. Avoid FIELD_PREP

FIELD_PREP macros does a few sanity checks as part of field updates.
Use a simple method to update fields as it helps improve throughput

Signed-off-by: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -161,18 +161,14 @@ int ath12k_dp_tx_direct(struct ath12k_vi
 		goto fail_unmap_dma;
 	}
 
-	tcl_desc.buf_addr_info.info0 = FIELD_PREP(BUFFER_ADDR_INFO0_ADDR, skb_cb->paddr);
-	tcl_desc.buf_addr_info.info1 = FIELD_PREP(BUFFER_ADDR_INFO1_ADDR,
-			((uint64_t)skb_cb->paddr >> HAL_ADDR_MSB_REG_SHIFT));
-	tcl_desc.buf_addr_info.info1 |=
-		FIELD_PREP(BUFFER_ADDR_INFO1_SW_COOKIE, tx_desc->desc_id);
-	tcl_desc.info0 = FIELD_PREP(HAL_TCL_DATA_CMD_INFO0_DESC_TYPE,
-				    HAL_TCL_DESC_TYPE_BUFFER) | arvif->desc.info0;
+	tcl_desc.buf_addr_info.info0 = (u32)skb_cb->paddr;
+	tcl_desc.buf_addr_info.info1 = (tx_desc->desc_id << 12);
+	tcl_desc.info0 =  arvif->desc.info0;
 	tcl_desc.info1 =  arvif->desc.info1;
 
+	tcl_desc.info2 =  skb->len;
 	if (likely(skb->ip_summed == CHECKSUM_PARTIAL))
 		tcl_desc.info2 |= TX_IP_CHECKSUM;
-	tcl_desc.info2 |= FIELD_PREP(HAL_TCL_DATA_CMD_INFO2_DATA_LEN, skb->len);
 	tcl_desc.info3 = arvif->desc.info3;
 	tcl_desc.info4 =  arvif->desc.info4;
 	memcpy(hal_tcl_desc, &tcl_desc, sizeof(tcl_desc));
@@ -369,24 +365,16 @@ int ath12k_dp_tx(struct ath12k *ar, stru
 
 	tcl_cmd = (struct hal_tcl_data_cmd *)hal_tcl_desc;
 
-	tcl_cmd->buf_addr_info.info0 =
-		FIELD_PREP(BUFFER_ADDR_INFO0_ADDR, ti.paddr);
-	tcl_cmd->buf_addr_info.info1 =
-		FIELD_PREP(BUFFER_ADDR_INFO1_ADDR,
-			   ((uint64_t)ti.paddr >> HAL_ADDR_MSB_REG_SHIFT));
-	tcl_cmd->buf_addr_info.info1 |=
-		FIELD_PREP(BUFFER_ADDR_INFO1_SW_COOKIE, ti.desc_id);
-
-	tcl_cmd->info0 = FIELD_PREP(HAL_TCL_DATA_CMD_INFO0_DESC_TYPE,
-				    ti.type) | arvif->desc.info0;
-	tcl_cmd->info1 = FIELD_PREP(HAL_TCL_DATA_CMD_INFO1_CMD_NUM,
-				    ti.meta_data_flags);
-	tcl_cmd->info2 = ti.flags0 |
-		FIELD_PREP(HAL_TCL_DATA_CMD_INFO2_DATA_LEN, ti.data_len);
+	tcl_cmd->buf_addr_info.info0 = (u32)ti.paddr;
+	/* TODO: Copy the upper 8 bits here */
+	tcl_cmd->buf_addr_info.info1 = (ti.desc_id << 12);
+	tcl_cmd->info0 = (ti.type << 1) | arvif->desc.info0;
+	tcl_cmd->info1 = ti.meta_data_flags << 16;
+	tcl_cmd->info2 = ti.flags0 | ti.data_len;
 	tcl_cmd->info3 = arvif->desc.info3;
 	tcl_cmd->info4 = arvif->desc.info4;
 	tcl_cmd->info5 = 0;
-
+	dsb(st);
 	ath12k_hal_srng_access_umac_src_ring_end_nolock(ab->mem, tcl_ring);
 
 	ath12k_dbg_dump(ab, ATH12K_DBG_DP_TX, NULL, "dp tx msdu: ",
